source/MaterialXGenShader/ShaderStage.h

```diff
@@ -0,0 +1,303 @@
+    VariableBlock& getConstantBlock();
+
+    /// Return the constant variable block.
+    const VariableBlock& getConstantBlock() const;
+
+    /// Return a map of all uniform blocks.
+    const VariableBlockMap& getUniformBlocks() const
+    {
+        return _uniforms;
+    }
+
+    /// Return a map of all input blocks.
+    const VariableBlockMap& getInputBlocks() const
+    {
+        return _inputs;
+    }
+
+    /// Return a map of all output blocks.
+    const VariableBlockMap& getOutputBlocks() const
+    {
+        return _outputs;
+    }
+ 
+protected:
+    /// Start a new scope using the given bracket type.
+    void beginScope(Brackets brackets = Brackets::BRACES);
+
+    /// End the current scope.
+    void endScope(bool semicolon = false, bool newline = true);
+
+    /// Start a new line.
+    void beginLine();
+
+    /// End the current line.
+    void endLine(bool semicolon = true);
+
+    /// Add a newline character.
+    void newLine();
+
+    /// Add a string.
+    void addString(const string& str);
+
+    /// Add a single line of code, optionally appening a semi-colon.
+    void addLine(const string& str, bool semicolon = true);
+
+    /// Add a single line code comment.
+    void addComment(const string& str);
+
+    /// Add a block of code.
+    void addBlock(const string& str, GenContext& context);
+
+    /// Add the contents of an include file. Making sure it is 
+    /// only included once for the shader stage.
+    void addInclude(const string& file, GenContext& context);
+
+    /// Add a value.
+    template<typename T>
+    void addValue(const T& value)
+    {
+        std::stringstream str;
+        str << value;
+        _code += str.str();
+    }
+
+    /// Add the function definition for a node.
+    void addFunctionDefinition(const ShaderNode& node, GenContext& context);
+
+private:
+    /// Name of the stage
+    const string _name;
+
+    /// Syntax for the type of shader to generate.
+    ConstSyntaxPtr _syntax;
+
+    /// Current indentation level.
+    int _indentations;
+
+    /// Current scope.
+    std::queue<Brackets> _scopes;
+
+    /// Set of include files that has been included.
+    std::set<string> _includes;
+
+    /// Set of hash ID's for functions that has been defined.
+    std::set<size_t> _definedFunctions;
+
+    /// Block holding constant variables for this stage.
+    VariableBlock _constants;
+
+    /// Map of blocks holding uniform variables for this stage.
+    VariableBlockMap _uniforms;
+
+    /// Map of blocks holding input variables for this stage.
+    VariableBlockMap _inputs;
+
+    /// Map of blocks holding output variables for this stage.
+    VariableBlockMap _outputs;
+
+    /// Resulting source code for this stage.
+    string _code;
+
+    friend class ShaderGenerator;
+};
+
+using ShaderStagePtr = std::shared_ptr<ShaderStage>;
+
+/// Utility function for adding a new shader port to a uniform block.
+inline ShaderPort* addStageUniform(const string& block, 
```

Comment:
Shader variable creation can be done through direct calls on the VariableBlock class, but for convenience the following utility functions can be used for less verbose syntax.